### PR TITLE
bpo-45923: Add `RESUME_QUICK`

### PR DESCRIPTION
**Pull request by brandtbucher**
Also, clean up `JUMP_ABSOLUTE`/`JUMP_ABSOLUTE_QUICK` to reduce code duplication.

1% faster:

<details>

```
Slower (18):
- nqueens: 85.1 ms +- 1.6 ms -> 86.8 ms +- 0.8 ms: 1.02x slower
- pickle: 10.1 us +- 0.2 us -> 10.3 us +- 0.3 us: 1.02x slower
- regex_dna: 231 ms +- 1 ms -> 233 ms +- 1 ms: 1.01x slower
- sympy_expand: 493 ms +- 3 ms -> 498 ms +- 6 ms: 1.01x slower
- deltablue: 4.06 ms +- 0.06 ms -> 4.09 ms +- 0.07 ms: 1.01x slower
- json_loads: 27.0 us +- 0.3 us -> 27.2 us +- 0.5 us: 1.01x slower
- meteor_contest: 104 ms +- 0 ms -> 105 ms +- 1 ms: 1.01x slower
- xml_etree_generate: 78.5 ms +- 0.5 ms -> 79.0 ms +- 0.5 ms: 1.01x slower
- chameleon: 6.86 ms +- 0.05 ms -> 6.89 ms +- 0.07 ms: 1.01x slower
- tornado_http: 97.7 ms +- 1.2 ms -> 98.2 ms +- 1.3 ms: 1.01x slower
- sympy_sum: 166 ms +- 2 ms -> 167 ms +- 2 ms: 1.00x slower
- xml_etree_process: 55.9 ms +- 0.5 ms -> 56.2 ms +- 0.6 ms: 1.00x slower
- sympy_str: 295 ms +- 3 ms -> 296 ms +- 3 ms: 1.00x slower
- regex_v8: 25.2 ms +- 0.2 ms -> 25.3 ms +- 0.1 ms: 1.00x slower
- dulwich_log: 65.6 ms +- 0.3 ms -> 65.8 ms +- 0.4 ms: 1.00x slower
- sympy_integrate: 21.0 ms +- 0.1 ms -> 21.0 ms +- 0.2 ms: 1.00x slower
- python_startup_no_site: 6.00 ms +- 0.00 ms -> 6.01 ms +- 0.00 ms: 1.00x slower
- python_startup: 8.69 ms +- 0.01 ms -> 8.70 ms +- 0.01 ms: 1.00x slower

Faster (31):
- scimark_sparse_mat_mult: 5.02 ms +- 0.02 ms -> 4.80 ms +- 0.09 ms: 1.04x faster
- logging_silent: 110 ns +- 1 ns -> 106 ns +- 2 ns: 1.04x faster
- telco: 6.65 ms +- 0.25 ms -> 6.40 ms +- 0.14 ms: 1.04x faster
- scimark_fft: 342 ms +- 2 ms -> 330 ms +- 2 ms: 1.04x faster
- pickle_list: 4.53 us +- 0.07 us -> 4.38 us +- 0.06 us: 1.04x faster
- unpickle_list: 5.27 us +- 0.08 us -> 5.09 us +- 0.04 us: 1.03x faster
- richards: 50.5 ms +- 1.3 ms -> 48.9 ms +- 1.3 ms: 1.03x faster
- mako: 10.6 ms +- 0.1 ms -> 10.3 ms +- 0.1 ms: 1.03x faster
- pyflate: 460 ms +- 3 ms -> 448 ms +- 8 ms: 1.03x faster
- crypto_pyaes: 86.4 ms +- 1.2 ms -> 84.3 ms +- 0.9 ms: 1.02x faster
- unpack_sequence: 46.1 ns +- 0.5 ns -> 45.1 ns +- 0.7 ns: 1.02x faster
- nbody: 96.3 ms +- 1.3 ms -> 94.3 ms +- 2.3 ms: 1.02x faster
- unpickle: 14.2 us +- 0.1 us -> 13.9 us +- 0.2 us: 1.02x faster
- float: 77.5 ms +- 1.1 ms -> 76.2 ms +- 0.9 ms: 1.02x faster
- pickle_dict: 28.5 us +- 0.1 us -> 28.0 us +- 0.1 us: 1.02x faster
- django_template: 35.2 ms +- 0.4 ms -> 34.6 ms +- 0.3 ms: 1.02x faster
- scimark_monte_carlo: 71.7 ms +- 0.9 ms -> 70.6 ms +- 1.2 ms: 1.01x faster
- go: 144 ms +- 2 ms -> 142 ms +- 1 ms: 1.01x faster
- scimark_lu: 118 ms +- 4 ms -> 116 ms +- 2 ms: 1.01x faster
- pidigits: 194 ms +- 0 ms -> 192 ms +- 0 ms: 1.01x faster
- pycparser: 1.18 sec +- 0.02 sec -> 1.17 sec +- 0.01 sec: 1.01x faster
- scimark_sor: 125 ms +- 1 ms -> 124 ms +- 2 ms: 1.01x faster
- spectral_norm: 103 ms +- 1 ms -> 101 ms +- 3 ms: 1.01x faster
- logging_format: 6.03 us +- 0.10 us -> 5.96 us +- 0.09 us: 1.01x faster
- json_dumps: 12.3 ms +- 0.1 ms -> 12.2 ms +- 0.1 ms: 1.01x faster
- logging_simple: 5.49 us +- 0.08 us -> 5.43 us +- 0.08 us: 1.01x faster
- raytrace: 317 ms +- 6 ms -> 314 ms +- 6 ms: 1.01x faster
- chaos: 73.5 ms +- 1.4 ms -> 73.0 ms +- 0.6 ms: 1.01x faster
- xml_etree_iterparse: 105 ms +- 1 ms -> 105 ms +- 1 ms: 1.01x faster
- hexiom: 6.61 ms +- 0.04 ms -> 6.58 ms +- 0.04 ms: 1.00x faster
- unpickle_pure_python: 247 us +- 2 us -> 246 us +- 2 us: 1.00x faster

Benchmark hidden because not significant (13): 2to3, fannkuch, html5lib, json, pathlib, pickle_pure_python, regex_compile, regex_effbot, sqlalchemy_declarative, sqlalchemy_imperative, sqlite_synth, thrift, xml_etree_parse

Geometric mean: 1.01x faster
```

</details>

https://bugs.python.org/issue45923

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -1745,6 +1742,13 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
                 first_instr = frame->f_code->co_firstinstr;
                 next_instr = first_instr + nexti;
             }
+            JUMP_TO_INSTRUCTION(RESUME_QUICK);
+        }
+
+        TARGET(RESUME_QUICK) {
```
Comment (markshannon) [inline Python/ceval.c:1748]: Could you name for what it does, rather than how we got to it?
Maybe `CHECK_EVAL_BREAKER`?

File: Python/ceval.c [hunk 2]
```diff
@@ -1745,6 +1742,13 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+            PREDICTED(RESUME_QUICK);
+            assert(tstate->cframe == &cframe);
+            assert(frame == cframe.current_frame);
             frame->f_state = FRAME_EXECUTING;
             if (_Py_atomic_load_relaxed(eval_breaker) && oparg < 2) {
```
Comment (markshannon) [inline Python/ceval.c:1753]: The specializer should handle the `oparg < 2`.

File: Python/specialize.c [hunk 3]
```diff
@@ -405,6 +405,9 @@ optimize(SpecializedCacheOrInstruction *quickened, int len)
                 case JUMP_ABSOLUTE:
                     instructions[i] = _Py_MAKECODEUNIT(JUMP_ABSOLUTE_QUICK, oparg);
                     break;
+                case RESUME:
+                    instructions[i] = _Py_MAKECODEUNIT(RESUME_QUICK, oparg);
```
Comment (markshannon) [inline Python/specialize.c:409]: Handle the `oparg < 2` condition here
```
instructions[i] = _Py_MAKECODEUNIT(oparg < 2 ? CHECK_EVAL_BREAKER : NOP);
```

**Review by markshannon — CHANGES_REQUESTED** [review]
The `oparg < 2` check can be done when quickening.

**Comment by markshannon** [comment]
I note that https://github.com/python/cpython/pull/31240 also shows a ~1% speedup.
I wonder, is this a slow common base build, or two independent speedups?

File: Python/specialize.c (hunk 3 above)
Comment (brandtbucher) [reply Python/specialize.c:409]: Is `NOP` legal here? Wouldn't we still need to set `frame->f_state = FRAME_EXECUTING`?

**Comment by brandtbucher** [comment]
> I note that #31240 also shows a ~1% speedup. I wonder, is this a slow common base build, or two independent speedups?

I'm pretty sure they're independent, since even though the `main` base changeset is the same, they use two different builds and two different `pyperformance` runs.

All four combinations of PR branches and `main` runs show a 1% improvement:

- `resume-quick` vs `main` (from this PR)
- `resume-quick` vs `main` (from #31240)
- `unpack-sequence` vs `main` (from this PR)
- `unpack-sequence` vs `main` (from #31240)

I know the UK benchmarking machine is a bit more stable than the US one, though, so if you want to verify those numbers I won't be offended. 🙃

File: Python/specialize.c (hunk 3 above)
Comment (markshannon) [reply Python/specialize.c:409]: I forgot about `f_state`. We should get rid of it, but that's a separate issue.

**Comment by markshannon** [comment]
OK, merging this. We can quicken `RESUME (oparg <2)` to `NOP` once `f_state` has been removed.

**Pull request merged**